astropy/utils/masked/function_helpers.py

```diff
@@ -1055,7 +1055,13 @@ def array2string(
 
 @dispatched_function
 def array_str(a, max_line_width=None, precision=None, suppress_small=None):
-    # Override to avoid special treatment of array scalars.
+    # Override to change special treatment of array scalars, since the numpy
+    # code turns the masked array scalar into a regular array scalar.
+    # Note that MaskedFormat takes a regular format function, which expects
+    # a numpy scalar, but np.array_str expects an array, hence the indirection.
+    if a.shape == () and a.dtype.names is None:
+        return MaskedFormat(lambda x: np.array_str(np.asarray(x)))(a)
```

Comment:
Lambda functions are created every time. This function doesn't refer to any locals, so I think it should be declared separately.

Reply:
Yes, makes sense. Done.